### ISSUE TITLE
Expanded glTF support

### ISSUE THREAD

**Issue by jskinn** [pyvista/pyvista#1469]
As part of my usage of pyvista, I found the need to use 3D models in glTF format (see https://www.khronos.org/gltf/).
The glTF format has better support for physically-based rendering (PBR) materials than other model formats, and is useful for web display. It is the format expected by [three.js](https://threejs.org/) and [model-viewer](https://modelviewer.dev/), amongst others.

So far, I have two particular use cases:
- Exporting a plot as glTF. Currently the only supported formats are OBJ (which doesn't support materials well), and VTKJS, which I haven't experimented with, but cannot be integrated with other formats/models as far as I can tell. glTF export could be useful to get models into other 3D applications as well (such as Blender).
- Import and display of glTF models. Again, mostly for the PBR materials over other formats.

VTK already includes glTF support, so for file IO, it is mostly a matter of adding the existing implementation.
As far as I can tell, `vtkGLTFReader` can be added to the list of READERS at the top of `pyvista.utilities.fileio`, and everything will just work.
Similarly, I copied `export_obj` in `BasePlotter` and replaced `vtkOBJExporter` with `vtkGLTFExporter`, and everything worked.
Most of what else needs to happen is testing.

I have a few remaining issues/thoughts:
- Coordinate frames on export. glTF uses a different coordinate frame to pyvista/VTK with Y up and Z forward (see [here](https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#coordinate-system-and-units)). The VTK exporter does not handle this, so exported models wind up on their side. I think it would be more intuitive if the export could handle this, but I couldn't find a way to apply a general transformation to an entire plot.
- Support for more export formats. While it would be straightforward to create an `export_gltf` alongside `export_obj`, it might be cleaner to support a single generic export function, and allow the format to be specified from either the file name or an override, similar to the `read` API in `fileio`.
- Display of models with embedded PBR materials is an issue. While the geometry imports easily, `plotter.add_mesh` does not seem designed to read detailed colour information from the mesh.  From some experimentation, material properties from the file are stored in named arrays on the `PolyData` object, which must be read into the correct values. A model may contain multiple materials with different properties, which are expressed as multiple `PolyData` objects inside a `MultiData` object. I've managed to get a glTF model with PBR materials to correctly display, but it required copying a significant chunk of `BasePlotter.add_mesh` to unpack the `MultiData` object and read the relevant properties for each `PolyData`. I'm also concerned that the particular named arrays that material properties are stored in depend on the model format the data was imported from, so there might not be a generic way to handle reading material data without some custom fiddling during import. This might be big enough to become a separate issue in the future.

**Comment by akaszynski** [comment]
> VTK already includes glTF support, so for file IO, it is mostly a matter of adding the existing implementation.
As far as I can tell, vtkGLTFReader can be added to the list of READERS at the top of pyvista.utilities.fileio, and everything will just work.
Similarly, I copied export_obj in BasePlotter and replaced vtkOBJExporter with vtkGLTFExporter, and everything worked.
Most of what else needs to happen is testing.

Looks like this can be easily added then.  Let me work on a PR for it.

> Coordinate frames on export. glTF uses a different coordinate frame to pyvista/VTK with Y up and Z forward (see here). The VTK exporter does not handle this, so exported models wind up on their side. I think it would be more intuitive if the export could handle this, but I couldn't find a way to apply a general transformation to an entire plot.

I'm not sure this is necessary.  I think that if you intend on your scene being rendered in three.js, you should just have your model oriented in that manner and adjust the camera position in VTK.  It's no that the two are using different coordinate systems (they're both cartesian), just the default camera positions are different.

> Support for more export formats. While it would be straightforward to create an export_gltf alongside export_obj, it might be cleaner to support a single generic export function, and allow the format to be specified from either the file name or an override, similar to the read API in fileio.

This is a good approach for it, that way we don't have to add a method for each export.

> Display of models with embedded PBR materials is an issue. While the geometry imports easily, plotter.add_mesh does not seem designed to read detailed colour information from the mesh. From some experimentation, material properties from the file are stored in named arrays on the PolyData object, which must be read into the correct values. A model may contain multiple materials with different properties, which are expressed as multiple PolyData objects inside a MultiData object. I've managed to get a glTF model with PBR materials to correctly display, but it required copying a significant chunk of BasePlotter.add_mesh to unpack the MultiData object and read the relevant properties for each PolyData. I'm also concerned that the particular named arrays that material properties are stored in depend on the model format the data was imported from, so there might not be a generic way to handle reading material data without some custom fiddling during import. This might be big enough to become a separate issue in the future.

This would indeed have to be a separate issue/PR, and I'm going to need help with this one.

**Comment by akaszynski** [comment]
This is really cool.  PR should be out today.

```py
filename = '/home/alex/.local/share/pyvista/examples/DamagedHelmet.gltf'

from vtkmodules.vtkIOImport import vtkGLTFImporter
importer = vtkGLTFImporter()
importer.SetFileName(filename)

import pyvista
pl = pyvista.Plotter()
importer.SetRenderWindow(pl.ren_win)
importer.Update()
pl.show()
```

![sc](https://user-images.githubusercontent.com/11981631/124401666-e0159900-dce7-11eb-9416-10fb5a5c7b7a.png)

**Comment by jskinn** [comment]
I too had a go at a pull request for it, which in retrospect I should have mentioned, see: https://github.com/jskinn/pyvista/tree/feat/gltf-support



>> Coordinate frames on export. glTF uses a different coordinate frame to pyvista/VTK with Y up and Z forward (see here). The VTK exporter does not handle this, so exported models wind up on their side. I think it would be more intuitive if the export could handle this, but I couldn't find a way to apply a general transformation to an entire plot.
> 
> I'm not sure this is necessary. I think that if you intend on your scene being rendered in three.js, you should just have your model oriented in that manner and adjust the camera position in VTK. It's no that the two are using different coordinate systems (they're both cartesian), just the default camera positions are different.

I agree, this is not a necessary feature, and the workflow you describe would be a good way of content-authoring for web. I just figured that if there was an easy way to rotate all objects before export, it would sidestep all the "but why is my plot sideways" questions. If there is not, it's not worth the trouble.

> `import pyvista`
> `pl = pyvista.Plotter()`
> `importer.SetRenderWindow(pl.ren_win)`
> `importer.Update()`
> `pl.show()`
Huh, I did not know you could display models this way. I tried to do it via `reader` and then `add_mesh`, which is where I ran in to difficulties. What is the intended workflow for for loading a model from file, and then showing it?

**Comment by akaszynski** [comment]
> Huh, I did not know you could display models this way. I tried to do it via reader and then add_mesh, which is where I ran in to difficulties. What is the intended workflow for for loading a model from file, and then showing it?

That's the intended workflow since a glTF file contains the "scene" and not the mesh.  Idea being that you just import the scene which includes lighting, textures, meshes, etc.

> I too had a go at a pull request for it, which in retrospect I should have mentioned, see: https://github.com/jskinn/pyvista/tree/feat/gltf-support

Looked at it and I think my PR uses a similar approach.  I'll add this from your PR:
```
        if not hasattr(self, "ren_win"):
            raise RuntimeError("This plotter must still have a render window open.")
```
Since it doesn't check for it. 

When you get a chance, please take a look at #1489.

**Comment by SugarRayLua** [comment]
Dear @akaszynski ,
Thanks for creating pyvista and describing a way to import glTF files. I'm a novice programmer but share some of the same interest in using glTF files as @jskinn. I'm especially interested in creating 3D scenes and then exporting them as a single file .html and sharing those .html files with colleagues, friends and families cross-platform and without the individuals that I send the files needing to know how to find and use a viewer to open up gLTF files (instead, they can just open the single file .html). Now, with the information you provided above, I can import such gLTF files along with their proper PBR textures.

Two questions:
1. Since this issue was closed in 2021, is there any simpler method to import gLTF files with their associated textures then the above code:
```
from vtkmodules.vtkIOImport import vtkGLTFImporter
importer = vtkGLTFImporter()
importer.SetFileName(filename)

import pyvista
pl = pyvista.Plotter()
importer.SetRenderWindow(pl.ren_win)
importer.Update()
pl.show()
```

I confirmed that the code above still works; I was hoping that something simpler like pyvista.read() could import gLTF/glb but pyvista.read seems to only import the meshes and not the texture (or at least not render the texture)

2. Is there a way that inside a Jupyter notebook with pyvista loaded one could export a loaded gLTF seen to a single file .html and have that single file .html also include and render the texture in the gLTF/glb? I really like pyvista's export to html button in the jupyter notebook scene, but it seems that when I use that button to export the imported gLTF/glb mesh/scene, I only get the mesh/scene without the texture.

I'm also an R user, and pyvista reminds me of the R rgl package but somewhat easier to use and perhaps with some more capabilities. 

Thanks! :-)
